### ISSUE TITLE
Disable explicit cast for widening conversions

### ISSUE THREAD

**Issue by arthrp** [riok/mapperly#1638]
**Is your feature request related to a problem? Please describe.**
Currently there's always an explicit cast when casting one numeric type to another. However, it's not needed for widening conversions! Consider the following example:

```csharp
public class DtoA
{
    public byte Count { get; set; }
}

public class DtoB
{
    public int Count { get; set; }
}

[Mapper]
public static partial class Mapper
{
    public static partial DtoB Map(DtoA a);
}
```

It produces this mapper:
```csharp
    public static partial class Mapper
    {
        [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "4.1.1.0")]
        public static partial global::MapperlyExample.DtoB Map(global::MapperlyExample.DtoA a)
        {
            var target = new global::MapperlyExample.DtoB();
            target.Count = (int)a.Count;
            return target;
        }
    }
```

The cast is redundant. Some might also view the explicit cast as a code smell, since if you change the `DtoA` Count to be long:

```csharp
public class DtoA
{
    public long Count { get; set; }
}
```

the mapping code stays the same - but now you have potential data loss as it's a narrowing conversion!

**Describe the solution you'd like**
Either:

* Change default behavior not to produce explicit casts in mappers on numeric types widening conversions
OR
* Introduce parameter to control that behavior i.e. `[Mapper(NoExplicitWideningCast = true)]`

**Comment by TonEnfer** [comment]
Hi @arthrp 
It seems that what you are asking for can be achieved to some extent with the following setup: `[Mapper(EnabledConversions = MappingConversionType.All & ~MappingConversionType.ExplicitCast)]`. [Docs](https://mapperly.riok.app/docs/api/riok.mapperly.abstractions.mappingconversiontype/#fields)
Maybe this solution will suit you?

**Comment by arthrp** [comment]
Thanks, but applying `[Mapper(EnabledConversions = MappingConversionType.All & ~MappingConversionType.ExplicitCast)]` actually produces identical mapping code to the one shown above (with explicit cast). Also same when allowing only implicit casts `EnabledConversions = MappingConversionType.ImplicitCast` - looks like a bug? :)

**Comment by TonEnfer** [comment]
Indeed, the above code with `byte` to `int` mapping still produces code with cast, but the following code does not generate cast, instead it produces error RMG007:
```
public class DtoA
{
    public long Count { get; set; }
}

public class DtoB
{
    public int Count { get; set; }
}

[Mapper(EnabledConversions = MappingConversionType.All & ~MappingConversionType.ExplicitCast)]
public static partial class AnimalCloningExtensions
{
    public static partial DtoB Map(DtoA source); //<- RMG007 Could not map member ConsoleApp1.DtoA.Count of type long to ConsoleApp1.DtoB.Count of type int
}
```
Are you getting a different result? If so, I think you should provide information about your environment and code sample (see the bug issue template) and wait for the developers to respond

**Comment by latonz** [comment]
After a few more thoughts on this one I think we should disable the explicit cast conversion by default with the next breaking change. Happy to accept PRs 😊 
General idea:
* Introduce a new `MappingConversionType.Default = All & ~ExplicitCast`
* Replace all usages of `MappingConversionType.All` with `MappingConversionType.Default` (including in the docs)
* Update the tests
* Add the new `Default` conversion type to the docs and update the description of the `ExplicitCast` conversion type
* Create a migration guide or add an entry to the existing one in `docs/docs/breaking-changes`
  * enable for a single mapping => user-implemented mapping
  * enable explicit cast for all types => `EnabledConversions = MappingConversionType.All`

**Comment by zchung** [comment]
> After a few more thoughts on this one I think we should disable the explicit cast conversion by default with the next breaking change. Happy to accept PRs 😊 General idea:
> 
> * Introduce a new `MappingConversionType.Default = All & ~ExplicitCast`
> * Replace all usages of `MappingConversionType.All` with `MappingConversionType.Default` (including in the docs)
> * Update the tests
> * Add the new `Default` conversion type to the docs and update the description of the `ExplicitCast` conversion type
> * Create a migration guide or add an entry to the existing one in `docs/docs/breaking-changes`
>   
>   * enable for a single mapping => user-implemented mapping
>   * enable explicit cast for all types => `EnabledConversions = MappingConversionType.All`

@latonz Hi!

I'm new to being a contributor, thought I would have a go at adding this feature.

Thanks in advance for any feedback. :)
